### DISCUSSION TITLE
problem starting server

### DISCUSSION THREAD

**Discussion by JulesArcher** [JuliaPluto/PlutoSliderServer.jl#83] [Issues] [answered]
Hi,
First of all, many thanks to the development team for the amazing work you have done to create Pluto and the server version! 
I am trying to set up my own private linux VPS server and am almost there! 

I have been following the procedure in the PlutoSliderServer.jl Read.me file. I have made a few changes to serve my notebooks from a local folder rather than from a github folder, but otherwise have attempted to follow the steps where relevant.

Before starting the server set-up:
1) I have successfully set up a nginx reverse-proxy with SSL on Ubuntu server 22.04 LTS
2) Verified that I can manually serve Pluto notebooks from a folder to a secure https://... location at the julia REPL 

Now for the server procedure:
1) Julia installed and can run as root (which julia says /usr/local/bin/julia)
2) I skipped this step, as I will serve from a local folder instead
3) Created the /etc/systemd/system/pluto-server.service.sh shell script, and verified that running manually at the command line **successfully launches the  server**
4)Created the /usr/local/bin/pluto-slider-server.sh start-up script
5) Set appropriate permissions via:
sudo chmod 744 /usr/local/bin/pluto-slider-server.sh
sudo chmod 664 /etc/systemd/system/pluto-server.service

6) Started the server via:
sudo systemctl daemon-reload
sudo systemctl start pluto-server
sudo systemctl enable pluto-server

The initial launch shows initial launch success via systemctl -l status pluto-server, however after a few seconds and re-running, it shows failure to complete due to an error related to running julia itself. 

From the logs (sudo journalctl -u pluto-server), the failure appears to be due to the following:

/usr/local/bin/pluto-slider-server.sh: line 3: julia: command not found

Any ideas as to what could be the reason that the julia command itself is not found when starting the server via systemctl but works when manually running the shell script to launch the server??

Thanks!

Jules

**Comment by rikhuijzer** [comment, marked as answer]
Does it work if you explicitly write `/usr/local/bin/julia` to call the `julia` binary in `/usr/local/bin/pluto-slider-server.sh`?

**Reply by JulesArcher** [reply]
Hi @rikhuijzer,

Thanks for the reply and suggestion! I gave this a try; after modifying the server shell script and re-doing step#5, I still get a server status error as follows for the same reason as before (julia command not found):

Process: 3803 ExecStart=/usr/local/bin/pluto-slider-server.sh (code=exited, status=1/FAILURE)

It's as if the systemctl execution is running in some 'root mode' state that doesn't 'see' the julia executable path. I also did the following (per instructions) to set up a link to the julia executable, and modified my .bashrc to point to the link file in /usr/local/bin/julia.  
sudo ln -s `pwd`/julia-$JULIA_VERSION/bin/julia /usr/local/bin/julia

Jules

**Reply by rikhuijzer** [reply]
Seems like a systemd issue and I don't know much about systemd. According to https://stackoverflow.com/a/42552705/5056635 "A systemd service runs with a clean environment, so it is not clobbered by any random environment the user happens to be in when starting a service.".

From that quote, I take that you probably have to explicitly make various paths and environment variables available so that the process spawned by systemd is allowed to look at them. This is probably for security reasons as nos on StackOverflow points out.

**Reply by JulesArcher** [reply]
@rikhuijzer, thanks again for the reply and some clues to follow-up with. Since I have been able to launch the server manually by running the shell script, this issue as you say is more likely a systemd/linux issue and not a Pluto server-specific issue. I'll keep trying, and of course follow-up whatever fix I might find. 

Thanks for the help!

Jules

**Reply by JulesArcher** [reply]
Success! 

systemd was running with root as user; I modified the pluto-server.service file to add the following two lines to run with a different user=username, group=admin (you can find out these settings by executing id username):

User=username
Group=admin

After re-starting the server via systemd, everything works fine now!

**Reply by fonsp** [reply]
Thanks for the report @JulesArcher! Can you be sure to [mark the answer](https://docs.github.com/en/discussions/collaborating-with-your-community-using-discussions/participating-in-a-discussion#marking-a-comment-as-an-answer)? Is there anything we could improve in our documentation to prevent this problem in the future? Consider [editing the documentation yourself](https://docs.github.com/en/repositories/working-with-files/managing-files/editing-files).
